usb/ehci/ehci.c

```diff
@@ -36,166 +36,271 @@
+{
+	last->hw->next = QTD_PTR_INVALID;
+	last->hw->token |= QTD_IOC;
+
+	/* No qtds linked */
+	if (qh->lastQtd == NULL)
+		qh->hw->nextQtd = QTD_PTR(first);
+	else
+		qh->lastQtd->next = QTD_PTR(first);
 	ehci_memDmb();
 
-	prevlast->next.terminate = 0;
+	qh->lastQtd = last->hw;
 }
 
 
-static void ehci_continue(struct qh_node *qhnode, struct qtd *last)
+static void ehci_continue(ehci_t *ehci, ehci_qh_t *qh, ehci_qtd_t *last)
 {
-	struct qh *qh = qhnode->qh;
+	mutexLock(ehci->asyncLock);
+	/* Queue for this qh is empty */
+	if (qh->lastQtd == last->hw) {
+		qh->lastQtd = NULL;
+		qh->hw->nextQtd = QTD_PTR_INVALID;
+	}
 
-	if (qhnode->last == last) {
-		qhnode->last = &qh->transferOverlay;
-		qhnode->last->next = last->next;
+	/* The qh is halted after error, move it forward and cleanup the status */
+	if (!(qh->hw->token & QTD_ACTIVE) && (qh->hw->current == QTD_PTR(last)) &&
+			(qh->hw->token & QTD_ERRMASK)) {
+		qh->hw->nextQtd = last->hw->next;
+		qh->hw->token &= ~0x7e;
 	}
-	else if (qh->transferOverlay.active == 0 && qh->currentQtd.pointer == va2pa(last) >> 5) {
-		qh->transferOverlay.next = last->next;
+	ehci_memDmb();
+	mutexUnlock(ehci->asyncLock);
+}
+
+
+static ehci_qtd_t *ehci_qtdGet(ehci_t *ehci)
+{
+	ehci_qtd_t *qtd;
+
+	mutexLock(ehci->asyncLock);
+	qtd = ehci->qtdPool;
+
+	if (qtd != NULL) {
+		LIST_REMOVE(&ehci->qtdPool, qtd);
+		ehci->nqtds--;
 	}
+	mutexUnlock(ehci->asyncLock);
+
+	return qtd;
 }
 
 
-static struct qtd *ehci_allocQtd(int token, size_t maxpacksz, char *buffer, size_t *size, int datax)
+static void ehci_qtdsPut(ehci_t *ehci, ehci_qtd_t **head)
 {
-	struct qtd *qtd = NULL;
+	ehci_qtd_t *q, *t;
+
+	mutexLock(ehci->asyncLock);
+	while ((q = *head) != NULL) {
+		LIST_REMOVE(head, q);
+
+		if (ehci->nqtds >= EHCI_MAX_QTD_POOL && ehci->nqtds != 0) {
+			t = ehci->qtdPool;
+			LIST_REMOVE(&ehci->qtdPool, t);
+			usb_free((void *)t->hw, sizeof(*t->hw));
+			free(t);
+			ehci->nqtds--;
+		}
+		LIST_ADD(&ehci->qtdPool, q);
+		ehci->nqtds++;
+	}
+	mutexUnlock(ehci->asyncLock);
+}
+
+
+static ehci_qtd_t *ehci_qtdAlloc(ehci_t *ehci, int pid, size_t maxpacksz, char *data, size_t *size, int datax)
+{
+	ehci_qtd_t *qtd;
 	size_t bytes = 0;
 	int i, offs;
 
-	if ((qtd = usb_alloc(sizeof(struct qtd))) == NULL)
-		return NULL;
+	/* Try to reuse a qtd */
+	if ((qtd = ehci_qtdGet(ehci)) == NULL) {
+		if ((qtd = malloc(sizeof(ehci_qtd_t))) == NULL)
+			return NULL;
+
+		if ((qtd->hw = usb_alloc(sizeof(struct qtd))) == NULL)
+			return NULL;
+	}
+
+	qtd->hw->token = (datax << 31) | (pid << 8) | (EHCI_TRANS_ERRORS << 10) | QTD_ACTIVE;
 
-	qtd->dt = datax;
-	qtd->pid = token;
+	qtd->hw->next = QTD_PTR_INVALID;
+	qtd->hw->altnext = QTD_PTR_INVALID;
 
-	qtd->next.terminate = 1;
-	qtd->altNext.terminate = 1;
-	qtd->active = 1;
-	qtd->errorCounter = 3;
+	if (data != NULL) {
+		qtd->hw->buf[0] = (uintptr_t)va2pa(data) & (EHCI_PAGE_SIZE - 1);
+		qtd->hw->buf[0] |= (uintptr_t)va2pa(data) & ~0xfff;
```

Comment:
EHCI_PAGE_SIZE - 1 is most likely equal to 0xfff, so this could be simplified to qtd->hw->buf[0] = (uintptr_t)va2pa(data); Anyway, one syscall less:
```suggestion
		qtd->hw->buf[0] = (uintptr_t)va2pa(data) & ((EHCI_PAGE_SIZE - 1) | ~0xfff);
```

Reply:
Also, I think the old code (>> 12) was about _PAGE_SIZE